### PR TITLE
make losses > binary where possible

### PR DESCRIPTION
**Pull request by olliethomas**

### CODE REVIEW

File: bolts/models/erm.py [hunk 1]
```diff
@@ -124,8 +124,8 @@ def test_step(self, batch: DataBatch, batch_idx: int) -> Dict[str, Tensor]:
     def training_step(self, batch: DataBatch, batch_idx: int) -> Tensor:
         logits = self.forward(batch.x)
         loss = self._get_loss(logits, batch)
-        target = batch.y.view(-1, 1).long()
-        acc = self.train_acc(logits >= 0, target)
+        target = batch.y.view(-1).long()
+        acc = self.train_acc(logits.softmax(-1), target)
```
Comment (MylesBartlett) [inline bolts/models/erm.py:128]: Can you not use argmax here?

File: bolts/losses/loss.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+class ClfL1(nn.L1Loss):
+    def __init__(
+        self, out_dim: int, size_average=None, reduce=None, reduction: str = 'mean'
+    ) -> None:
+        super().__init__(size_average=size_average, reduce=reduce, reduction=reduction)
+        self.out_dim = out_dim
+
+    def forward(self, input: Tensor, target: Tensor) -> Tensor:
+        _target = target.view(-1, self.out_dim).float()
```
Comment (MylesBartlett) [inline bolts/losses/loss.py:18]: Why is outdim needed here? Could you not just do `target = target.viewas(input)` instead? The inputs to the L1 loss have to have the same shape.

File: bolts/models/laftr.py [hunk 3]
```diff
@@ -129,29 +133,27 @@ def _inference_step(self, batch: DataBatch, stage: str) -> Dict[str, Tensor]:
...
     def _loss_adv(self, s_pred: Tensor, batch: DataBatch) -> Tensor:
         # For Demographic Parity, for EqOpp is a different loss term.
-        s_target = batch.s.view(-1, 1)
-        y_target = batch.y.view(-1, 1)
         if self.fairness is FairnessType.DP:
-            s0 = self._adv_clf_loss(s_pred[batch.s == 0], s_target[batch.s == 0])
-            s1 = self._adv_clf_loss(s_pred[batch.s == 1], s_target[batch.s == 1])
+            s0 = self._adv_clf_loss(s_pred[batch.s == 0], batch.s[batch.s == 0])
+            s1 = self._adv_clf_loss(s_pred[batch.s == 1], batch.s[batch.s == 1])
             loss = (s0 + s1) / 2
```
Comment (olliethomas) [inline bolts/models/laftr.py:139]: I feel like this could be a 1 liner

Comment (olliethomas) [reply bolts/models/laftr.py:139]: @MylesBartlett any ideas?

Comment (MylesBartlett) [reply bolts/models/laftr.py:139]: err just do `loss(s_pred, s)?`

File: bolts/models/laftr.py [hunk 4]
```diff
@@ -129,29 +133,27 @@ def _inference_step(self, batch: DataBatch, stage: str) -> Dict[str, Tensor]:
...
         elif self.fairness is FairnessType.EO:
             loss = torch.tensor(0.0).to(self.device)
             for s, y in itertools.product([0, 1], repeat=2):
-                if len(batch.s[(batch.s == s) & (y_target == y)]) > 0:
+                if len(batch.s[(batch.s == s) & (batch.y == y)]) > 0:
                     loss += self._adv_clf_loss(
-                        s_pred[(s_target == s) & (y_target == y)],
-                        batch.s[(s_target == s) & (y_target == y)],
+                        s_pred[(batch.s == s) & (batch.y == y)],
+                        batch.s[(batch.s == s) & (batch.y == y)],
```
Comment (thomkeh) [inline bolts/models/laftr.py:146]: for some reason I can't submit this as a suggestion
```python
                mask = (batch.s == s) & (batch.y == y)
                if len(batch.s[mask]) > 0:
                     loss += self._adv_clf_loss(
                         s_pred[mask], batch.s[mask]
```

File: bolts/models/laftr.py (hunk 3 above)
Comment (olliethomas) [reply bolts/models/laftr.py:139]: no, because I need the average of each group, which is different

File: bolts/models/laftr.py (hunk 4 above)
Comment (olliethomas) [reply bolts/models/laftr.py:146]: 👍

File: bolts/models/laftr.py (hunk 3 above)
Comment (MylesBartlett) [reply bolts/models/laftr.py:139]: you could do it by reshaping and reducing

File: bolts/models/laftr.py [hunk 5]
```diff
@@ -129,39 +133,32 @@ def _inference_step(self, batch: DataBatch, stage: str) -> Dict[str, Tensor]:
...
             loss = (s0 + s1) / 2
         elif self.fairness is FairnessType.EO:
             loss = torch.tensor(0.0).to(self.device)
             for s, y in itertools.product([0, 1], repeat=2):
-                if len(batch.s[(batch.s == s) & (y_target == y)]) > 0:
-                    loss += self._adv_clf_loss(
-                        s_pred[(s_target == s) & (y_target == y)],
-                        batch.s[(s_target == s) & (y_target == y)],
-                    )
+                if len(batch.s[(batch.s == s) & (batch.y == y)]) > 0:
```
Comment (thomkeh) [inline bolts/models/laftr.py:143]: This should be equivalent, right?
```suggestion
                if ((batch.s == s) & (batch.y == y)).sum() > 0:
```

**Review by MylesBartlett — APPROVED** [review]

File: bolts/models/laftr.py [hunk 6]
```diff
@@ -129,39 +133,33 @@ def _inference_step(self, batch: DataBatch, stage: str) -> Dict[str, Tensor]:
...
             for s, y in itertools.product([0, 1], repeat=2):
-                if len(batch.s[(batch.s == s) & (y_target == y)]) > 0:
-                    loss += self._adv_clf_loss(
-                        s_pred[(s_target == s) & (y_target == y)],
-                        batch.s[(s_target == s) & (y_target == y)],
-                    )
-            loss = 2 - loss
+                mask = (batch.s == s) & (batch.y == y)
+                unweighted_loss[mask] /= mask.sum()
+            loss = 2 - unweighted_loss.sum() / len(list(itertools.product([0, 1], repeat=2)))
```
Comment (MylesBartlett) [inline bolts/models/laftr.py:147]: Can you make this product iterable a local variable instead of recomputing it?

File: bolts/models/laftr.py [hunk 7]
```diff
@@ -129,39 +133,33 @@ def _inference_step(self, batch: DataBatch, stage: str) -> Dict[str, Tensor]:
...
-        y_target = batch.y.view(-1, 1)
         if self.fairness is FairnessType.DP:
-            s0 = self._adv_clf_loss(s_pred[batch.s == 0], s_target[batch.s == 0])
-            s1 = self._adv_clf_loss(s_pred[batch.s == 1], s_target[batch.s == 1])
-            loss = (s0 + s1) / 2
+            losses = self._adv_clf_loss(s_pred, batch.s)
+            for s in (0, 1):
+                mask = batch.s == s
+                losses[mask] /= mask.sum()
+            loss = 1 - losses.sum() / len((0, 1))
```
Comment (MylesBartlett) [inline bolts/models/laftr.py:141]: just hardcode the denominator as 2

File: bolts/models/laftr.py [hunk 8]
```diff
@@ -129,39 +133,33 @@ def _inference_step(self, batch: DataBatch, stage: str) -> Dict[str, Tensor]:
...
-                if len(batch.s[(s_target == s) & (y_target == 1)]) > 0:
-                    loss += self._adv_clf_loss(
-                        s_pred[(s_target == s) & (y_target == 1)],
-                        batch.s[(s_target == s) & (y_target == 1)],
-                    )
-            loss = 2 - loss
+                mask = (batch.s == s) & (batch.y == 1)
+                unweighted_loss[mask] /= mask.sum()
+            unweighted_loss[batch.y == 0] *= 0.0
+            loss = 2 - unweighted_loss.sum() / len((0, 1))
```
Comment (MylesBartlett) [inline bolts/models/laftr.py:155]:
```suggestion
            loss = 2 - unweighted_loss.sum() / 2
```
